File: plugin/src/main/java/com/google/tsunami/plugin/payload/Payload.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.tsunami.plugin.payload;

import com.google.common.flogger.GoogleLogger;
import com.google.protobuf.ByteString;
import com.google.tsunami.proto.PayloadAttributes;
import com.google.tsunami.proto.PayloadGeneratorConfig;
import java.util.Optional;

/** Type returned by {@link PayloadGenerator} to be used in detectors. */
public class Payload {

  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
  private final String payload;
  private final Validator validator;
  private final PayloadAttributes attributes;
  private final PayloadGeneratorConfig config;

  public Payload(
      String payload,
      Validator validator,
      PayloadAttributes attributes,
      PayloadGeneratorConfig config) {
    this.payload = payload;
    this.validator = validator;
    this.attributes = attributes;
    this.config = config;
  }

  /**
   * Get the string representation of the payload.
   *
   * @return the actual payload string
   */
  public final String getPayload() {
    logger.atInfo().log(
        "%s generated payload `%s`, %s use the callback server",
        this.config, this.payload, this.attributes.getUsesCallbackServer() ? "does" : "does not");
    return this.payload;
  }

  /**
   * Checks if the supplied payload was executed based on a given input e.g. a reflective RCE.
   *
   * @param input - a UTF-8 encoded string
   * @return whether this payload is executed on the scan target.
   */
  public final boolean checkIfExecuted(String input) {
    return this.validator.isExecuted(Optional.of(ByteString.copyFromUtf8(input)));
  }

  /**
   * Checks if the supplied payload was executed based on a given input e.g. a reflective RCE.
   *
   * @param input - a sequence of bytes in the {@link ByteString} format.
   * @return whether this payload is executed on the scan target.
   */
  public final boolean checkIfExecuted(ByteString input) {
    return this.validator.isExecuted(Optional.of(input));
  }

  /**
   * Checks if the supplied payload was executed based on a given input e.g. a reflective RCE.
   *
   * @param input - an optional sequence of bytes in the {@link ByteString} format.
   * @return whether this payload is executed on the scan target.
   */
  public final boolean checkIfExecuted(Optional<ByteString> input) {
    return this.validator.isExecuted(input);
  }

  /**
   * Checks if the supplied payload was executed without supplying an input e.g. validation against
   * the callback server does not require input.
   *
   * @return whether this payload is executed on the scan target.
   */
  public final boolean checkIfExecuted() {
    return this.validator.isExecuted(Optional.empty());
  }

  /**
   * Get additional attributes about this payload.
   *
   * @return the {@link PayloadAttributes} about this payload
   */
  public final PayloadAttributes getPayloadAttributes() {
    return this.attributes;
  }
}

File: plugin_server/py/plugin/payload/payload.py
"""Payload is the type returned by the Payload Generator."""

from typing import Optional

from absl import logging

from tsunami.plugin_server.py.plugin.payload.validator import Validator
from tsunami.proto import payload_generator_pb2


class Payload:
  """Out-of-bound payload to be sent to the scan target."""

  def __init__(
      self,
      payload: str,
      validator: Validator,
      attributes: payload_generator_pb2.PayloadAttributes,
      config: payload_generator_pb2.PayloadGeneratorConfig,
  ):
    self.payload = payload
    self.validator = validator
    self.attributes = attributes
    self.config = config

  def get_payload(self) -> str:
    """Get the string representation of the payload.

    Returns:
      The payload string
    """
    logging.info(
        '%s generated payload `%s`, %s use the callback server',
        self.config,
        self.payload,
        'does' if self.attributes.uses_callback_server else 'does not',
    )
    return self.payload

  def check_if_executed(
      self, payload_data: Optional[str | bytes] = None
  ) -> bool:
    """Check if the payload was executed on the scan target.

    Args:
      payload_data: Optional string or bytees representation of the payload to
        be verified.

    Returns:
      Whether the payload was executed.
    """
    payload = (
        payload_data.encode('utf-8')
        if isinstance(payload_data, str)
        else payload_data
    )
    return self.validator.is_executed(payload)

  def get_payload_attributes(self) -> payload_generator_pb2.PayloadAttributes:
    """Get the payload attributes.

    Returns:
      Details of the payload.
    """
    return self.attributes
